Sprint 16

DarkRavenn opened the pull request.

DarkRavenn committed:
Реализован онбординг

DarkRavenn committed:
Рефакторинг:
- Перемещение изображений в Assets
- Добавлена иконка приложения

DarkRavenn committed:
Добавлен экран с выбором категории

DarkRavenn committed:
Изменена логика определения был ли просмотрен онбординг

andrew-kuzn-yndx commented on Tracker/Application/Screens/Onboarding/OnboardingPageFactory.swift:
```diff
@@ -0,0 +1,94 @@
+        if #available(iOS 14.0, *) {
+            // Используется для iOS 14.0 и новее
+            button = UIButton(primaryAction: UIAction { _ in
+                action() // Выполняем переданное замыкание
```
Старайся не оставлять такие комментарии, когда заливаешь изменения в основную ветку проекта. 
Я понимаю, что такие комментарии помогают тебе писать код, но после разработки фичи их можно убирать. В репозиторий должен попадать только чистый код) 
Конечно, это не касается критически важных комментариев

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/TrackerCreation/Category/CategoryTableCell.swift:
```diff
@@ -0,0 +1,52 @@
+            } else if indexPath.row == 0 {
+                // первая строчка (скругляем сверху)
+                layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner]
+            } else if indexPath.row == (superview as! UITableView).numberOfRows(inSection: indexPath.section) - 1 {
```
В этих строках есть force операция - force cast, она как и force unwrap является нежелательной и надо стараться такого кода избегать. Можно переписать без их использования, переместив в мягкий каст в if: if let a = a as? 
Всегда когда есть возможность написать безопасную конструкцию – отдавай предпочтение ей. Так твоя работа будет выигрышнее смотреться)

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/TrackerCreation/Category/CategoryViewController.swift:
```diff
@@ -0,0 +1,192 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! CategoryTableCell
```
Тут тоже можно от force cast избавиться)

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/TrackerCreation/Category/CategoryViewController.swift:
```diff
@@ -0,0 +1,192 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as! CategoryTableCell
+        cell.textLabel?.text = vm.categories[indexPath.row]
```
Старайся не работать со свойствами ячейки и вообще класса вот так напрямую из другого класса. 
По хорошему - это сделать в ячейке публичную функцию, например configCell с аргументами на вход, и уже в ней менять нужные свойства. 
Свойства класса нужно всегда стараться закрывать извне, и работать с ними только изнутри

andrew-kuzn-yndx commented on Tracker/Application/Screens/Trackers/TrackerCreation/Category/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,117 @@
+        self.updateDoneButtonState()
+    }
+    
+    @objc func createButtonTapped() {
```
Не забывай, что @objc func **всегда** должны быть приватными

DarkRavenn committed:
Правки по ревью:
- Добавлено закрытие клавиатуры тапом за пределом клавиатуры при создании категории
- В дни когда нет трекеров отображается заглушка "Что будем отслеживать?"
- Исправлен баг после добавления новой категории не отображались разделители
- Поправлены цвета
- Убрал force cast

DarkRavenn committed:
Правки по ревью:
- убрано прямое взаимодействие с свойствами ячейки

DarkRavenn committed:
Исправлено отображение заглушки "Что будем отслеживать?"

DarkRavenn committed:
Удалён неиспользуемый код
